0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣🔟

# 귀무가설

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats
%matplotlib inline
plt.rc('figure', figsize=(10, 6))
from matplotlib import rcParams
rcParams['font.family'] = 'Noto Sans KR'
rcParams['font.size'] = 10
rcParams['axes.unicode_minus'] = False

In [9]:
#  1. 가설검정 단계
    # H0 : 대한민국 남성의 평균 몸무게는 70킬로그램이다(차이가 없다)
    # H1 : 대한민국 남성의 평균 몸무게는 70킬로그램이 아니다 (차이가 있다)


#  2. 통계량 산출 - 데이터 수집
np.random.seed(123)
weight = np.random.uniform(40, 100, 100)

test_result = stats.ttest_1samp(weight, 70)
print(f'평균 : {np.mean(weight)}')
print(f'p-value : {test_result.pvalue}')

#  3. 기각/채택 여부 판단
print(f'검정결과 : {'귀무가설 기각' if test_result.pvalue<0.05 else '귀무가설 채택'}')
#  4. 결정
if test_result.pvalue<0.05:
    print('대한민국 남성의 평균 몸무게는 70kg이 아니다')
else :
    print('대한민국 남성의 평균 몸무게는 70kg이다')    

평균 : 70.08573280308887
p-value : 0.953913826658112
검정결과 : 귀무가설 채택
대한민국 남성의 평균 몸무게는 70kg이다


# 2️⃣ T검정

### 예제 1: 평균 비교 - 일표본(단표본) t검정

**문제:** 하루 커피 섭취량이 평균 2.5잔이라고 알려져 있다. 실제 설문조사에서 2.8잔이 나왔다. 과연 차이가 통계적으로 유의한가?

In [4]:
# 표본 데이터 (30명)
coffee_data = [2.9, 2.4, 2.7, 3.0, 2.6, 2.8, 2.9, 3.1, 2.7, 2.6,
               2.8, 2.5, 3.2, 3.0, 2.4, 2.6, 2.7, 3.1, 2.8, 2.7,
               2.5, 2.9, 2.6, 2.8, 2.9, 2.5, 2.7, 2.8, 3.0, 2.9]

# 단일표본 t-검정
t_stat, p_value = stats.ttest_1samp(coffee_data, popmean=2.5) #popmean 비교할 평균 

print(f"t값 = {t_stat:.3f}, p값 = {p_value:.10f}")

t값 = 7.031, p값 = 0.0000000986


t값은 표본과 모집단 평균 간 차이가 표준오차의 7배라는 통계량
-> 표본 평균이 모집단 2.5에서 7*표준오차 만큼 떨어져 있다.
-> 매우 큰 차이를 보인다.   

이 통계량은 p-value < 0.05 이므로 통계적으로 유의한 차이가 있다고 할 수 있습니다.


기존 귀무가설의 내용가 차이가 없다(귀무가설 선택)


### 예제 2: 두 그룹 비교 - 이표본(독립표본) t검정

**문제:**
H0 :  A와 B 수업의 시험 평균 점수가 차이가 없다
H1 :  A와 B 수업의 시험 평균 점수 차이가 있다

In [ ]:
group_A = [78, 85, 82, 88, 76, 81, 79, 84, 82, 80]
group_B = [72, 75, 78, 74, 71, 77, 70, 76, 75, 73]

t_stat, p_value = stats.ttest_ind(group_A, group_B, equal_var=False) #ind 함수 / equal_var :다르다고 주장 
print(f"t값 = {t_stat:.3f}, p값 = {p_value:.10f}")

t값 = 5.331, p값 = 0.0000604698


H0 :  A와 B 수업의 시험 평균 점수가 차이가 없다   
H1 :  A와 B 수업의 시험 평균 점수 차이가 있다

통계량: 5.331, 표준오차의 5배 정도 차이가 난다 
유의확률: 0.00006 < 0.05 보다 작다 -> 귀무가설 기각 

귀무가설을 기각/채택 : 귀무가설을 기각한다.
결론 : 두 그룹의 평균은 유의미하게 차이가 있다.



### 예제 3: 전후의 비교 -  대응 표본 T 검정  

In [7]:
# 대응 표본 t-검정 예제: 약물 투여 전후 혈압 비교
before = [120, 122, 143, 130, 135, 118, 127, 140, 132, 125]
after = [115, 117, 138, 125, 129, 112, 120, 135, 126, 119]

t_stat, p_value = stats.ttest_rel(before, after)
print(f"t값 = {t_stat:.3f}, p값 = {p_value:.8f}")

t값 = 25.327, p값 = 0.00000000


H0 :  약물 투여의 혈압 변동은 차이가 없다 
H1 :  약물 투여와 혈압 변동은 차이가 있다

통계량: 25.32, 표준오차의 25배 정도 차이가 난다 
유의확률: 0.00006 < 0.05 보다 작다 -> 귀무가설 기각 

귀무가설을 기각/채택 : 귀무가설을 기각한다.
결론 : 두 그룹의 평균은 유의미하게 차이가 있다.



### ### 예제 4: 범주형 변수 - 카이제곱 검정

**문제:** 성별에 따라 커피를 더 선호하는지?

In [10]:
import numpy as np
obs = np.array([[40, 10],  # 남자 (좋아함, 안좋아함) 교차표
                [30, 20]]) # 여자

chi2, p, dof, expected = stats.chi2_contingency(obs)
print(f"chi2 = {chi2:.3f}, p값 = {p:.8f}")

chi2 = 3.857, p값 = 0.04953461


H0 :  성별에 따라 커피 선호도의 차이는 없다 . 
H1 :  성별에 따라 커피 선호도의 차이는 있다.

통계량: 3.857, 표준오차의 3배 정도 차이가 난다 
유의확률: 0.00006 < 0.05 보다 작다 -> 귀무가설 기각 

귀무가설을 기각/채택 : 귀무가설을 기각한다.
결론 : 두 그룹의 평균은 유의미하게 차이가 있다.



In [11]:
from scipy.stats import f_oneway

# 세 그룹의 점수
group1 = [80, 85, 90]
group2 = [70, 75, 78]
group3 = [60, 65, 62]

f_stat, p_value = f_oneway(group1, group2, group3)
print(f"F값: {f_stat:.2f}, p값: {p_value:.4f}")

F값: 24.28, p값: 0.0013


H0 :  세개 그룹의 평균은 차이가 없다. 같다   
H1 :  적어도 1개 그룹은 다르다 

통계량: 24, 그룹간 분산의 차이가 그룹내 분산보다 크다
유의확률: 0.00006 < 0.05 보다 작다 -> 귀무가설 기각 

귀무가설을 기각/채택 : 귀무가설을 기각한다.
결론 : 두 그룹의 평균은 유의미하게 차이가 있다.

